File: #163_Missing_Ranges/Solution.java
import java.util.*;

public class Solution {
	public List<String> findMissingRanges(int[] nums, int lower, int upper) {
		List<String> result = new LinkedList();
		if (upper < lower) return result;
		if (nums == null || nums.length == 0) {
			if (lower == upper) result.add(Integer.toString(lower));
			else result.add(Integer.toString(lower) + "->" + Integer.toString(upper));
			return result;
		}
		/* handle the first gap */
		if (lower < nums[0]) {
			int lo = lower; 
			int hi = nums[0] - 1;
			if (lo == hi) result.add(Integer.toString(lo));
			else result.add(Integer.toString(lo) + "->" + Integer.toString(hi));
		}
		/* handle the middle elements */
		for (int i = 1; i < nums.length; i++) {
			if (nums[i] - nums[i - 1] > 1) {
				int lo = nums[i - 1] + 1;
				int hi = nums[i] - 1;
				if (lo == hi) result.add(Integer.toString(lo));
				else result.add(Integer.toString(lo) + "->" + Integer.toString(hi));
			}
		}
		/* handle the last gap */
		if (upper > nums[nums.length - 1]) {
			int lo = nums[nums.length - 1] + 1; 
			int hi = upper;
			if (lo == hi) result.add(Integer.toString(lo));
			else result.add(Integer.toString(lo) + "->" + Integer.toString(hi));
		}
		return result;
	}
}

File: #163_Missing_Ranges/Solution.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} lower
    # @param {integer} upper
    # @return {string[]}
    def findMissingRanges(self, nums, lower, upper):
        if not nums:
        	if lower == upper:
        		return [str(lower)]
        	return [str(lower) + '->' + str(upper)]
        re = []
        if lower < nums[0]:
    		ub = nums[0] - 1
    		lb = lower
    		if ub == lb:
    			re.append(str(lb))
        	elif ub > lb:
        		re.append(str(lb) + '->' + str(ub))
        for x in range(len(nums) - 1):
    		ub = nums[x + 1] - 1
    		lb = nums[x] + 1
    		if ub == lb:
    			re.append(str(lb))
    		elif ub > lb:
    			re.append(str(lb) + '->' + str(ub))
    	if nums[len(nums) - 1] < upper:
    		ub = upper
    		lb = nums[len(nums) - 1] + 1
    		if ub == lb:
    			re.append(str(lb))
    		elif ub > lb:
    			re.append(str(lb) + '->' + str(ub))
    	return re

sol = Solution()
print sol.findMissingRanges([1, 2, 3, 50, 99], 0, 99)